### ISSUE TITLE
Adding __slots__ to a function-like Protocol breaks support for function assigment to the Protocol

### ISSUE THREAD

**Issue by HexDecimal** [microsoft/pyright#9878]
## Environment data

-   Pylance version: 2025.2.1
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): CPython 3.12.0

## Code Snippet

This example is valid under strict Mypy and at runtime:
```python
from typing import Protocol


class Foo(Protocol):

    __slots__ = ()

    def __call__(self) -> None:
        ...


def bar() -> None:
    pass


callback: Foo = bar  # UNEXPECTED Pylance(reportAssignmentType)
callback()
```

## Expected behavior

Should pass type checking without errors.

Type checking should ignore `__slots__` in a Protocol.

## Actual behavior

Because of `__slots__ = ()` Pylance gives the following error:
```
Type "() -> None" is not assignable to declared type "Foo"
  "function" is incompatible with protocol "Foo"
    "__call__" is not present
```

## Logs

Too long to paste.

**Comment by erictraut** [comment]
This case involves an unspecified (or at least underspecified) part of the typing spec. Refer to the [protocols chapter](https://typing.readthedocs.io/en/latest/spec/protocol.html) for details. The current spec does not provide exemptions for any symbols in a protocol definition. However, type checker maintainers have (by convention) chosen to exempt `__slots__` when doing protocol matching.

Pyright includes the `__slots__` exemption in its main protocol matching algorithm. However, I didn't extend this exemption to the code path that determines whether a protocol is a [callback protocol](https://typing.readthedocs.io/en/latest/spec/callables.html#callback-protocols). For consistency, this code path should likewise exempt `__slots__`.

This change will be included in the next release of pyright.

**Comment by HexDecimal** [comment]
I think you're right in this being underspecified. I'm mainly somewhat familiar with CPython's handling of protocols.

For reference and transparency I've once made a PR to solve a similar issue in Mypy https://github.com/python/mypy/pull/15490 which references other issues explaining the use-cases and motivation for [excluding these CPython special attributes](https://github.com/python/cpython/blob/00ec7818771903e3007928d191d1297cdb3b5277/Lib/typing.py#L1866-L1872), though the issue with Pylance seems to be specifically about assigning functions to function-like protocols because when I try similar things with normal classes I don't encounter this error.

Here's another example with the same error, showing that a function-like class passes while a true function fails:

```py
from __future__ import annotations

from typing import Protocol, Self, Type


class Foo(Protocol):
    def __new__(cls) -> Self:
        return super().__new__(cls)

    def __init__(self) -> None:
        pass

    def __class_getitem__(cls, key: object) -> Type[Foo]:
        return cls

    def __call__(self) -> None: ...


def bar() -> None:
    pass


class Bar:
    def __call__(self) -> None:
        pass


callback: Foo = bar  # UNEXPECTED Pylance(reportAssignmentType)
callback()

value: Foo = Bar()  # But this passes without errors as expected
```
I point this out for completeness. My own use-case was only for `__slots__`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.394.
